### ISSUE TITLE
Memory leak - Deatch never gets called

### ISSUE THREAD

**Issue by lukasf** [jlaanstra/Windows.UI.Interactivity#2]
Hi, the current versions leak memory. Detach is never called, so the AssociatedObject stays in memory forever, causing memory footprint to rise and rise when changing pages until out of memory crash. I think perhaps in Interactivity.cs you need to attach to Unloaded event and then call Detach. I will also try that myself.

Thanks for working on this project, I am glad to have Interactivity on Metro! Hopefully soon without memory leak so I can really start using it.

**Comment by lukasf** [comment]
Update: I fixed the memory leak in 1.0.2.0 with the following change in AttachableCollectionOfT:

public void Attach(FrameworkElement frameworkElement)
        {
            if (frameworkElement != this.AssociatedObject)
            {
                if (this.AssociatedObject != null)
                {
                    throw new InvalidOperationException();
                }
                if (!Windows.ApplicationModel.DesignMode.DesignModeEnabled) // TODO: Verificare qui
                {
                    this.AssociatedObject = frameworkElement;
                    this.AssociatedObject.Unloaded += AssociatedObject_Unloaded;
                }
                this.OnAttached();
            }
        }

```
    void AssociatedObject_Unloaded(object sender, RoutedEventArgs e)
    {
        this.AssociatedObject.Unloaded -= AssociatedObject_Unloaded;
        Detach();
    }
```

However, this does only work with 1.0.2.0. With 1.0.3.0 with the same change, the application still leaks. I am pretty sure that this is due to the new DataContextChangedDetector. My guess is: It dynamically creates a new dependency property with a static fixed reference to the behavior instance. This reference is never released. Generally, I think it is a bad idea to abuse the DependencyProperty system, adding more and more properties will slow it all down. 

I suggest you either roll back to the previous DataContext handling, or use a different approach. This one seems good to me, it also uses weak reference to avoid memory leaks: http://agsmith.wordpress.com/2008/04/07/propertydescriptor-addvaluechanged-alternative/

**Comment by jlaanstra** [comment]
I'll look ito the memory leak issue soon. The previous DataContext handling was buggier and had specific issues when used in popup's or other elements not directly connected to the RootVisual. The approach you suggest would be great if it was supported in the Windows Runtime, but it isn't and until it is supported, this is the only option to actually get change notifications.

**Comment by lukasf** [comment]
Hi Johan,

I got the change handler code working in WinRT with some small modifications. With this, I got 1.0.3.0 leak free. Please check the code below.

Kind Regards
Lukas

```
    public Task WaitForDataContext(FrameworkElement obj)
    {
        if(obj.DataContext != null)
        {
            this.Complete(obj.DataContext);
        }
        else
        {
            PropertyChangeNotifier notifier = new PropertyChangeNotifier(obj, "DataContext");
            notifier.ValueChanged += (s, e) =>
                {
                    if (obj.DataContext != null)
                    {
                        this.Complete(obj.DataContext);
                        notifier.Dispose();
                    }
                };
        }
        return tcs.Task;
    }
```

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Data;

namespace Windows.UI.Interactivity
{
    public sealed class PropertyChangeNotifier : DependencyObject, IDisposable
    {
        #region Member Variables
        private WeakReference _propertySource;
        #endregion // Member Variables

```
    #region Constructor
    public PropertyChangeNotifier(DependencyObject propertySource, string path)
        : this(propertySource, new PropertyPath(path))
    {
    }
    //public PropertyChangeNotifier(DependencyObject propertySource, DependencyProperty property)
    //    : this(propertySource, new PropertyPath(property))
    //{
    //}
    public PropertyChangeNotifier(DependencyObject propertySource, PropertyPath property)
    {
        if (null == propertySource)
            throw new ArgumentNullException("propertySource");
        if (null == property)
            throw new ArgumentNullException("property");
        this._propertySource = new WeakReference(propertySource);
        Binding binding = new Binding();
        binding.Path = property;
        binding.Mode = BindingMode.OneWay;
        binding.Source = propertySource;
        BindingOperations.SetBinding(this, ValueProperty, binding);
    }
    #endregion // Constructor

    #region PropertySource
    public DependencyObject PropertySource
    {
        get
        {
            try
            {
                // note, it is possible that accessing the target property
                // will result in an exception so i’ve wrapped this check
                // in a try catch
                return this._propertySource.IsAlive
                ? this._propertySource.Target as DependencyObject
                : null;
            }
            catch
            {
                return null;
            }
        }
    }
    #endregion // PropertySource

    #region Value
    /// <summary>
    /// Identifies the <see cref="Value"/> dependency property
    /// </summary>
    public static readonly DependencyProperty ValueProperty = DependencyProperty.Register("Value",
    typeof(object), typeof(PropertyChangeNotifier), new PropertyMetadata(null, new PropertyChangedCallback(OnPropertyChanged)));

    private static void OnPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
    {
        PropertyChangeNotifier notifier = (PropertyChangeNotifier)d;
        if (null != notifier.ValueChanged)
            notifier.ValueChanged(notifier, EventArgs.Empty);
    }

    /// <summary>
    /// Returns/sets the value of the property
    /// </summary>
    /// <seealso cref="ValueProperty"/>
    public object Value
    {
        get
        {
            return (object)this.GetValue(PropertyChangeNotifier.ValueProperty);
        }
        set
        {
            this.SetValue(PropertyChangeNotifier.ValueProperty, value);
        }
    }
    #endregion //Value

    #region Events
    public event EventHandler ValueChanged;
    #endregion // Events

    #region IDisposable Members
    public void Dispose()
    {
        ValueChanged = null; // needed to avoid re-raising event
        BindingOperations.SetBinding(this, ValueProperty, null);
    }
    #endregion
}
```

}

-------- Original-Nachricht --------

**Comment by jlaanstra** [comment]
Thnx Lucas!

Your initial issue is fixed in ac98c20ddbd617cc98b20bc08b911a9b3362174b. I solved the second issue in another way since the code you supplied does not work in WinRT. Right now there are only two attached properties used instead of a number linear to the number of behaviors. It looks a lot cleaner to me and also solves the second leaking issue.

Regards,
Johan
